subsys/net/lib/nrf_cloud/src/nrf_cloud_logs.c

```diff
@@ -0,0 +1,452 @@
+			       NRF_CLOUD_TOPIC_BULK : NRF_CLOUD_TOPIC_BIN
+	};
+	uint32_t stored;
+	uint8_t *base64_buf = NULL;
+
+	/* The bulk topic requires the multiple JSON messages to be placed in
+	 * a JSON array. Close the array then send it.
+	 */
+	if ((num_msgs != 0) && (log_format_current == LOG_OUTPUT_TEXT)) {
+		ring_buf_put(&log_nrf_cloud_rb, "]", 1);
+	}
+
+	stored = ring_buf_size_get(&log_nrf_cloud_rb);
+	output.data.len = ring_buf_get_claim(&log_nrf_cloud_rb,
+					     (uint8_t **)&output.data.ptr,
+					     stored);
+	if (output.data.len != stored) {
+		LOG_WRN("Capacity:%u, free:%u, stored:%u, claimed:%u",
+			ring_buf_capacity_get(&log_nrf_cloud_rb),
+			ring_buf_space_get(&log_nrf_cloud_rb),
+			stored, output.data.len);
+		stored = output.data.len;
+	}
+
+	if ((log_format_current == LOG_OUTPUT_DICT) &&
+	    IS_ENABLED(CONFIG_NRF_CLOUD_LOG_DICT_JSON)) {
+		/* Convert to base64 and wrap in JSON */
+		size_t olen;
+		size_t buflen;
+
+		/* Determine buffer size needed */
+		base64_encode(NULL, 0, &olen, output.data.ptr, output.data.len);
+		buflen = strlen(JSON_FMT1) + olen + strlen(JSON_FMT2);
+
+		base64_buf = k_calloc(buflen, 1);
+		if (base64_buf == NULL) {
+			err = -ENOMEM;
+			goto cleanup;
+		}
+
+		memcpy(base64_buf, JSON_FMT1, strlen(JSON_FMT1));
+		err = base64_encode(&base64_buf[strlen(JSON_FMT1)], olen, &olen,
+				    output.data.ptr, output.data.len);
+		if (err) {
+			goto cleanup;
+		}
+		memcpy(&base64_buf[strlen(JSON_FMT1) + olen], JSON_FMT2, strlen(JSON_FMT2));
+		output.data.ptr = base64_buf;
+		output.data.len = buflen - 1;
+	} /* Else send ring buffer contents as is -- JSON or raw binary. */
+
+	if (IS_ENABLED(CONFIG_NRF_CLOUD_MQTT)) {
+		err = nrf_cloud_send(&output);
+	} else if (IS_ENABLED(CONFIG_NRF_CLOUD_REST)) {
+		/* send to d2c topic, unless bulk is true, in which case d2c/bulk is used */
+		err = nrf_cloud_rest_send_device_message(context.rest_ctx,
+							 context.device_id,
+							 output.data.ptr, false, NULL);
+	} else {
+		err = -ENODEV;
+	}
+	if (!err) {
+		stats.lines_sent += num_msgs;
+		stats.bytes_sent += output.data.len;
+	}
+
+cleanup:
+	if (base64_buf) {
+		k_free(base64_buf);
+	}
+	ret = ring_buf_get_finish(&log_nrf_cloud_rb, stored);
+	ring_buf_reset(&log_nrf_cloud_rb);
+	num_msgs = 0;
+
+	if (ret) {
+		LOG_ERR("Error finishing ring buffer: %d", ret);
+		err = ret;
+	}
+	return err;
+}
+
+static int logger_out(uint8_t *buf, size_t size, void *ctx)
+{
+	ARG_UNUSED(ctx);
+	int err = 0;
+	struct nrf_cloud_data data;
+	size_t orig_size = size;
+	uint32_t stored;
+	int extra;
+
+	if (!size) {
+		return 0;
+	}
+
+	if (log_format_current == LOG_OUTPUT_TEXT) {
+		extra = 2;
+		if (context.src_name) {
+			int len = strlen(context.src_name);
+
+			/* We want to pass the log source as a separate field for cloud filtering.
+			 * Because the Zephyr logging subsystem prefixes the log text with
+			 * 'src_name: ', remove it before encoding the log.
+			 */
```

Comment:
The `cr_out_func` is for when the user enables `LOG_PRINTK` which uses raw logging.  If there is no `src_name` then this code isn't used because it isn't needed -- there is no prefix to the string to remove.

Reply:
> I think this depends on the logging mode pretty heavily, does this method account for that?
> 
> I could be mistaken, though, I'm not very familiar with how the logging system works -- Maybe the formatting applied to each back-end can be independent, and you can lock this backend to have a consistent format regardless of log mode

Not sure what the "this" in "this depends" is referring to exactly.

Reply:
`this` was meant to refer to the format the string being parsed is in, and the process of parsing it

Reply:
It appears that the source name I Attempt to remove comes from the same source of the name the Zephyr log_output.c uses in the log prefix string.  There are no Kconfigs to disable putting the source name in the log prefix string, other than CONFIG_LOG_MINIMAL, which is incompatible with this logger since it just prints directly to printk().